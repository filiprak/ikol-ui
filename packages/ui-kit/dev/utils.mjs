import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
