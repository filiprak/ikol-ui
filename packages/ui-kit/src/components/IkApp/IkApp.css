body {
    position: relative;
    background-color: var(--background-color);
    color: var(--on-background-color-1);
    font-family: "Open Sans", sans-serif;
    letter-spacing: 0;
    font-weight: 600;
    font-size: 13px;
}

::selection {
    background-color: var(--primary-color) !important;
    color: var(--on-primary-color) !important;
}

::-moz-selection {
    background-color: var(--primary-color) !important;
    color: var(--on-primary-color) !important;
}

::placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-moz-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-webkit-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

:-ms-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-ms-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

/* number input arrows */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

input[type=number] {
    -moz-appearance: textfield;
}

.ik-app {
    display: grid;
    align-content: start;
    position: relative;
    -webkit-backface-visibility: hidden;
    backface-visibility: hidden;
    min-height: 100%;
    max-width: 100%;
}

.ik-app--fill {
    width: 100%;
    height: 100%;
    align-content: normal;
}

.ik-app-bar {
    position: fixed;
    top: 0;
    left: 0;
    flex: 0 0 auto;
    display: flex;
    align-items: center;
    width: 100%;
    max-width: 100%;
    z-index: var(--z-index-app-bar);
    background-color: var(--background-color);
    border-bottom: 1px solid var(--surface-darken-color);
}

.ik-app-bar__back,
.ik-app-bar__burger {
    padding-left: 15px;
    font-size: 24px;
}

.ik-app-bar__title {
    max-width: 50%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    font-size: 18px;
    padding-left: 15px;
    min-width: 0;
}

.ik-app--mobile .ik-app-bar__title {
    max-width: 100%;
}

.ik-app-bar__content {
    flex: 1 1 auto;
    padding: 0 25px;
    min-width: 0;
}

.ik-app-content {
    display: grid;
    min-height: 0;
}

input::-webkit-contacts-auto-fill-button {
    visibility: hidden;
    display: none !important;
    pointer-events: none;
    position: absolute;
    right: 0;
}
