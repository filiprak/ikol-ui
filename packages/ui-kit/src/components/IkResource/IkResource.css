.ik-resource:not(.ik-resource--empty) {
    display: inline-flex;
    max-width: 100%;
    align-items: center;
    cursor: pointer;
}

.ik-resource__content {
    flex-grow: 1;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 100%;
    border-bottom: 1px solid transparent;
}

.ik-resource:hover > .ik-resource__content {
    border-bottom: 1px dotted;
}

.ik-resource__actions {
    flex-grow: 0;
    display: none;
}

.ik-resource__actions > .ik-resource__action {
    padding: 0 1px 0 5px;
}

.ik-resource:hover > .ik-resource__actions {
    display: inline-flex;
}

.ik-resource--no-wrap .ik-resource__content {
    white-space: nowrap;
  }
