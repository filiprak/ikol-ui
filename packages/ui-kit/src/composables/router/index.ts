import type { InjectionKey, Ref } from 'vue';
import { inject } from 'vue';
import { getCurrentInstance, logError } from '@ui/utils/core';
import { IkRouterManager } from '@ui/composables/router/IkRouterManager';
import type { IkRouteManager } from '@ui/composables/router/IkRouteManager';
import type { RouteRecordRaw } from 'vue-router';
import { createRouter as _createRouter, createWebHashHistory } from 'vue-router';

export const ROUTER_SYMBOL: InjectionKey<IkRouterManager> = Symbol.for('ik-router');

export interface IkRouterOptions {
    routes: Readonly<RouteRecordRaw[]>
}

export function createRouter(options: IkRouterOptions): IkRouterManager {
    return new IkRouterManager(
        _createRouter({
            history: createWebHashHistory(),
            routes: options.routes,
        })
    );
}

export function useRouter(): IkRouterManager {
    let router;

    if (getCurrentInstance('useRouter')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router;
}

export function useRoute(): Ref<IkRouteManager> {
    let router;

    if (getCurrentInstance('useRoute')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router.route;
}
