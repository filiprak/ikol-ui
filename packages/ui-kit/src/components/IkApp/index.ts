import IkApp from './IkApp.vue';
import IkAppContent from './IkAppContent.vue';
import IkAppBar from './IkAppBar.vue';

export interface IkAppContent extends InstanceType<typeof IkAppContent> { };
export interface IkAppBar extends InstanceType<typeof IkAppBar> { };
export interface IkApp extends InstanceType<typeof IkApp> { };

export { IkApp };
export { IkAppContent };
export { IkAppBar };
