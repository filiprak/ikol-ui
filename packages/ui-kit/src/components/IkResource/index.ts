import IkResource from './IkResource.vue';
export interface IkResource extends InstanceType<typeof IkResource> { };
export { IkResource };
