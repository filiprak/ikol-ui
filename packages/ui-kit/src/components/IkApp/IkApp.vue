<script lang="ts">
import '@ui/styles';
import './IkApp.css';
import { defineComponent, h, unref } from 'vue';
import { toCssUnits } from '@ui/utils/helpers';

import type { IkAppBar } from '@ui/components/IkApp';
import type { IkNavigation } from '@ui/components/IkNavigation';
import { IkModalRenderer } from '@ui/components/IkModal';
import { useDevice } from '@ui/composables/device';

export default defineComponent({
    provide() {
        return {
            app: this,
        };
    },
    props: {
        loading: {
            type: Boolean,
            default: false,
        },
        fill: {
            type: Boolean,
            default: false,
        },
    },
    data() {
        return {
            nav: null as IkNavigation | null,
            bar: null as IkAppBar | null,
        };
    },
    computed: {
        styles() {
            if (!unref(this.device.mobile)) {
                return {
                    paddingLeft: toCssUnits(this.nav_width),
                    paddingTop: toCssUnits(this.bar_height),
                };
            } else {
                return {
                    paddingTop: toCssUnits(this.bar_height),
                };
            }
        },
        nav_width() {
            if (this.nav) {
                return this.nav.mini_width;
            } else {
                return 0;
            }
        },
        bar_height() {
            if (this.bar) {
                return this.bar.height;
            } else {
                return 0;
            }
        },
    },
    methods: {
        registerBar(bar: IkAppBar) {
            this.bar = bar || null;
        },
        registerNav(nav: IkNavigation) {
            this.nav = nav || null;
        },
        unregisterBar() {
            this.bar = null;
        },
        unregisterNav() {
            this.nav = null;
        },
    },
    render() {
        let children = this.$slots.default?.() || [];

        if (this.loading) {
            children = [];
        }

        children.push(h(IkModalRenderer));

        return h('div', {
            class: {
                'ik-app': true,
                'ik-app--fill': this.fill,
                'ik-app--mobile': unref(this.device.mobile),
            },
            style: this.styles,
        }, children);
    },
    setup() { return { device: useDevice() }; },
});
</script>
