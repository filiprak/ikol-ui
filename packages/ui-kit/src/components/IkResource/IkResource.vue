<template>
    <div :class="root_class">
        <div v-if="!empty"
             class="ik-resource__content"
             @click.stop
             @mousedown.stop
             @mouseup.stop="default_handler">
            <slot></slot>
        </div>
        <div v-if="!empty"
             class="ik-resource__actions">
            <IkIcon v-if="copyable"
                    :icon="copy_icon"
                    class="ik-resource__action"
                    @click="onClickCopy" />
            <IkIcon v-if="uri"
                    icon="external-link:regular"
                    class="ik-resource__action"
                    @click="onClickUri" />
        </div>
        <slot v-else></slot>
    </div>
</template>

<script lang="ts">
import './IkResource.css';
import { defineComponent, computed, ref } from 'vue';
import { copyToClipboard } from '@ui/utils/helpers';
import { IkIcon } from '@ui/components/IkIcon';

export default defineComponent({
    name: 'IkResource',
    components: {
        IkIcon,
    },
    props: {
        value: {
            type: String,
            default: undefined,
        },
        copyable: {
            type: Boolean,
            default: false,
        },
        uri: {
            type: String,
            default: undefined,
        },
        target: {
            type: String,
            default: '_blank',
        },
        no_wrap: {
            type: Boolean,
            default: false,
        },
        email: {
            type: Boolean,
            default: false,
        },
        phone: {
            type: Boolean,
            default: false,
        },
    },
    setup(props, { slots }) {
        const copy_done = ref<boolean>(false);

        const root_class = computed(() => ({
            'ik-resource': true,
            'ik-resource--empty': empty.value,
            'ik-resource--no-wrap': props.no_wrap,
        }));

        const empty = computed(() => !props.uri && !props.copyable);

        const computed_uri = computed(() => {
            if (props.uri) {
                if (props.email) {
                    return `mailto:${props.uri}`;
                } else if (props.phone) {
                    return `tel:${props.uri}`;
                } else {
                    return props.uri;
                }
            }
            return null;
        });

        const copy_icon = computed(() => (copy_done.value ? 'check-double:regular' : 'copy:regular'));

        const default_handler = computed(() => (props.copyable ? onClickCopy : onClickUri));

        const onClickCopy = (event: MouseEvent) => {
            const to_copy = props.value ? String(props.value) : slots.default?.toString();
            if (to_copy) {
                copyToClipboard(to_copy.trim());
                if (!copy_done.value) {
                    copy_done.value = true;
                    setTimeout(() => {
                        copy_done.value = false;
                    }, 800);
                }
                event.preventDefault();
                event.stopPropagation();
            }
        };

        const onClickUri = (event: MouseEvent) => {
            if (computed_uri.value) {
                const mouse_click = event.button === 2 || event.button === 0;
                const mouse_middle = event.button === 1;
                const target = mouse_middle ? '_blank' : props.target;

                if (mouse_click || mouse_middle) {
                    window.open(computed_uri.value, target);
                }
            }

            event.preventDefault();
            event.stopPropagation();
        };

        return {
            root_class,
            empty,
            computed_uri,
            copy_icon,
            default_handler,
            onClickCopy,
            onClickUri,
        };
    },
});
</script>
