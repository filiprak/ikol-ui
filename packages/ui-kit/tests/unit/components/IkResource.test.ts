import type { RenderOptions } from '@ui/tests/tools/utils';
import { IkApp } from '@ui/components/IkApp';
import { renderShallow } from '@ui/tests/tools/utils';
import { IkResource } from '@ui/components/IkResource';
import { getRouterMock } from '@ui/tests/mocks/router';
import { mockSelectionAPI } from '@ui/tests/mocks/selectionAPI';
import { mockNavigationAPI } from '@ui/tests/mocks/navigationAPI';

describe('IkResource', () => {
    let router = getRouterMock();
    const getInstance = (options?: RenderOptions<typeof IkResource>) => renderShallow(IkResource, {
        global: {
            provide: { app: getAppInstance() },
        },
        router,
    }, options);
    const getAppInstance = () => renderShallow(IkApp).vm;

    beforeEach(() => {
        router = getRouterMock();
    });

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toStrictEqual(['ik-resource', 'ik-resource--empty']);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders copyable', () => {
        const wrapper = getInstance({ props: { copyable: true } });
        expect(wrapper.findAll('.ik-resource__action')).toHaveLength(1);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders as uri link', () => {
        const wrapper = getInstance({ props: { uri: 'http://amazon.com' } });
        expect(wrapper.findAll('.ik-resource__action')).toHaveLength(1);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders copyable and as uri link', () => {
        const wrapper = getInstance({ props: { copyable: true, uri: 'http://amazon.com' } });
        expect(wrapper.findAll('.ik-resource__action')).toHaveLength(2);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({ props: { copyable: true }, slots: { default: '<span>This is</span>Resource<b>Component</b>' } });
        expect(wrapper.find('.ik-resource__content').text()).toContain('This isResourceComponent');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('copies text value to clipboard', () => {
        mockSelectionAPI();

        const wrapper = getInstance({ props: { copyable: true }, slots: { default: 'resource' } });
        wrapper.find('.ik-resource__content').trigger('mouseup');
        expect(document.execCommand).toHaveBeenCalledWith('copy');
    });

    it('navigates to uri', () => {
        mockNavigationAPI();

        const wrapper = getInstance({ props: { uri: 'http:/unit.test.com?c=1&b=5#param=test' }, slots: { default: 'Uri test' } });
        wrapper.find('.ik-resource__content').trigger('mouseup', { button: 2 });
        expect(window.open).toHaveBeenCalledTimes(1);
        expect(window.open).toHaveBeenCalledWith('http:/unit.test.com?c=1&b=5#param=test', '_blank');

        wrapper.find('.ik-resource__content').trigger('mouseup', { button: 1 });
        expect(window.open).toHaveBeenCalledTimes(2);
        expect(window.open).toHaveBeenCalledWith('http:/unit.test.com?c=1&b=5#param=test', '_blank');

        wrapper.find('.ik-resource__content').trigger('mouseup', { button: 4 });
        expect(window.open).toHaveBeenCalledTimes(2);
    });

    it('opens blank tab if middle mouse button clicked', () => {
        mockNavigationAPI();

        const wrapper = getInstance({ props: { uri: 'http:/unit.test.com?c=1&b=5#param=test', target: '_self' }, slots: { default: 'Uri test' } });
        wrapper.find('.ik-resource__content').trigger('mouseup', { button: 2 });
        expect(window.open).toHaveBeenLastCalledWith('http:/unit.test.com?c=1&b=5#param=test', '_self');

        wrapper.find('.ik-resource__content').trigger('mouseup', { button: 1 });
        expect(window.open).toHaveBeenLastCalledWith('http:/unit.test.com?c=1&b=5#param=test', '_blank');
    });

    it('works with copy action button', () => {
        mockSelectionAPI();

        const wrapper = getInstance({ props: { copyable: true }, slots: { default: 'Uri test' } });
        wrapper.find('.ik-resource__action').trigger('click');
        expect(document.execCommand).toHaveBeenCalledWith('copy');
    });

    it('works with uri action button', () => {
        mockNavigationAPI();

        const wrapper = getInstance({ props: { uri: 'http:/unit.test.com?c=1&b=5#param=test', target: '_self' }, slots: { default: 'Uri test' } });
        wrapper.find('.ik-resource__action').trigger('click', { button: 0 });
        expect(window.open).toHaveBeenLastCalledWith('http:/unit.test.com?c=1&b=5#param=test', '_self');
    });

    it('renders in no wrap mode', () => {
        const wrapper = getInstance({ props: { uri: 'http:/unit.test.com?c=1&b=5#param=test', target: '_self', no_wrap: true }, slots: { default: 'Uri test' } });
        expect(wrapper.classes()).toContain('ik-resource--no-wrap');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('works with phone uri', () => {
        mockNavigationAPI();

        const wrapper = getInstance({ props: { phone: true, uri: '+48 123 213 213', target: '_self' }, slots: { default: 'Uri test' } });
        wrapper.find('.ik-resource__action').trigger('click', { button: 0 });
        expect(window.open).toHaveBeenLastCalledWith('tel:+48 123 213 213', '_self');
    });

    it('works with email uri', () => {
        mockNavigationAPI();

        const wrapper = getInstance({ props: { email: true, uri: 'adam.kowalski@testikol.pl', target: '_self' }, slots: { default: 'Uri test' } });
        wrapper.find('.ik-resource__action').trigger('click', { button: 0 });
        expect(window.open).toHaveBeenLastCalledWith('mailto:adam.kowalski@testikol.pl', '_self');
    });
});
