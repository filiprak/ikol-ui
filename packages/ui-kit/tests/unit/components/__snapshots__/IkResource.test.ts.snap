// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkResource renders as uri link 1`] = `
<div
  class="ik-resource"
>
  <div
    class="ik-resource__content"
  >
    
    
  </div>
  <div
    class="ik-resource__actions"
  >
    <!--v-if-->
    <ik-icon-stub
      circle="false"
      class="ik-resource__action"
      disabled="false"
      icon="external-link:regular"
      tag="i"
    />
  </div>
</div>
`;

exports[`IkResource renders copyable 1`] = `
<div
  class="ik-resource"
>
  <div
    class="ik-resource__content"
  >
    
    
  </div>
  <div
    class="ik-resource__actions"
  >
    <ik-icon-stub
      circle="false"
      class="ik-resource__action"
      disabled="false"
      icon="copy:regular"
      tag="i"
    />
    <!--v-if-->
  </div>
</div>
`;

exports[`IkResource renders copyable and as uri link 1`] = `
<div
  class="ik-resource"
>
  <div
    class="ik-resource__content"
  >
    
    
  </div>
  <div
    class="ik-resource__actions"
  >
    <ik-icon-stub
      circle="false"
      class="ik-resource__action"
      disabled="false"
      icon="copy:regular"
      tag="i"
    />
    <ik-icon-stub
      circle="false"
      class="ik-resource__action"
      disabled="false"
      icon="external-link:regular"
      tag="i"
    />
  </div>
</div>
`;

exports[`IkResource renders in no wrap mode 1`] = `
<div
  class="ik-resource ik-resource--no-wrap"
>
  <div
    class="ik-resource__content"
  >
    
    Uri test
    
  </div>
  <div
    class="ik-resource__actions"
  >
    <!--v-if-->
    <ik-icon-stub
      circle="false"
      class="ik-resource__action"
      disabled="false"
      icon="external-link:regular"
      tag="i"
    />
  </div>
</div>
`;

exports[`IkResource renders with correct default state 1`] = `
<div
  class="ik-resource ik-resource--empty"
>
  <!--v-if-->
  
  
</div>
`;

exports[`IkResource renders with default slot 1`] = `
<div
  class="ik-resource"
>
  <div
    class="ik-resource__content"
  >
    
    <span>
      This is
    </span>
    Resource
    <b>
      Component
    </b>
    
  </div>
  <div
    class="ik-resource__actions"
  >
    <ik-icon-stub
      circle="false"
      class="ik-resource__action"
      disabled="false"
      icon="copy:regular"
      tag="i"
    />
    <!--v-if-->
  </div>
</div>
`;
